### DISCUSSION TITLE
Best way to add nested model with relationship in Create

### DISCUSSION THREAD

**Discussion by jonyscathe** [fastapi/sqlmodel#1116] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI, HTTPException, Query
from sqlmodel import Field, Relationship, Session, SQLModel, create_engine, select
from typing import Annotated
from pydantic import AfterValidator



class TeamBase(SQLModel):
    name: str = Field(index=True)
    headquarters: str


class Team(TeamBase, table=True):
    id: int | None = Field(default=None, primary_key=True)

    heroes: list["Hero"] = Relationship(back_populates="team")


class TeamCreate(TeamBase):
    pass


class TeamPublic(TeamBase):
    id: int


class TeamUpdate(SQLModel):
    id: int | None = None
    name: str | None = None
    headquarters: str | None = None


class HeroBase(SQLModel):
    name: str = Field(index=True)
    secret_name: str
    age: int | None = Field(default=None, index=True)



class Hero(HeroBase, table=True):
    id: int | None = Field(default=None, primary_key=True)

    team_id: int | None = Field(default=None, foreign_key="team.id")
    team: Team | None = Relationship(back_populates="heroes")


class HeroPublic(HeroBase):
    id: int


class HeroCreate(HeroBase):
    team: (
        Annotated[
            TeamCreate,
            AfterValidator(Team.model_validate),
        ]
        | None
    ) = None    


class HeroUpdate(SQLModel):
    name: str | None = None
    secret_name: str | None = None
    age: int | None = None
    team_id: int | None = None


class HeroPublicWithTeam(HeroPublic):
    team: TeamPublic | None = None


class TeamPublicWithHeroes(TeamPublic):
    heroes: list[HeroPublic] = []


sqlite_file_name = "database.db"
sqlite_url = f"sqlite:///{sqlite_file_name}"

connect_args = {"check_same_thread": False}
engine = create_engine(sqlite_url, echo=True, connect_args=connect_args)


def create_db_and_tables():
    SQLModel.metadata.create_all(engine)


def get_session():
    with Session(engine) as session:
        yield session


app = FastAPI()


@app.on_event("startup")
def on_startup():
    create_db_and_tables()


@app.post("/heroes/", response_model=HeroPublic)
def create_hero(*, session: Session = Depends(get_session), hero: HeroCreate):
    db_hero = Hero.model_validate(hero)
    session.add(db_hero)
    session.commit()
    session.refresh(db_hero)
    return db_hero


@app.get("/heroes/", response_model=list[HeroPublic])
def read_heroes(
    *,
    session: Session = Depends(get_session),
    offset: int = 0,
    limit: int = Query(default=100, le=100),
):
    heroes = session.exec(select(Hero).offset(offset).limit(limit)).all()
    return heroes


@app.get("/heroes/{hero_id}", response_model=HeroPublicWithTeam)
def read_hero(*, session: Session = Depends(get_session), hero_id: int):
    hero = session.get(Hero, hero_id)
    if not hero:
        raise HTTPException(status_code=404, detail="Hero not found")
    return hero


@app.patch("/heroes/{hero_id}", response_model=HeroPublic)
def update_hero(
    *, session: Session = Depends(get_session), hero_id: int, hero: HeroUpdate
):
    db_hero = session.get(Hero, hero_id)
    if not db_hero:
        raise HTTPException(status_code=404, detail="Hero not found")
    hero_data = hero.model_dump(exclude_unset=True)
    for key, value in hero_data.items():
        setattr(db_hero, key, value)
    session.add(db_hero)
    session.commit()
    session.refresh(db_hero)
    return db_hero


@app.delete("/heroes/{hero_id}")
def delete_hero(*, session: Session = Depends(get_session), hero_id: int):
    hero = session.get(Hero, hero_id)
    if not hero:
        raise HTTPException(status_code=404, detail="Hero not found")
    session.delete(hero)
    session.commit()
    return {"ok": True}


@app.post("/teams/", response_model=TeamPublic)
def create_team(*, session: Session = Depends(get_session), team: TeamCreate):
    db_team = Team.model_validate(team)
    session.add(db_team)
    session.commit()
    session.refresh(db_team)
    return db_team


@app.get("/teams/", response_model=list[TeamPublic])
def read_teams(
    *,
    session: Session = Depends(get_session),
    offset: int = 0,
    limit: int = Query(default=100, le=100),
):
    teams = session.exec(select(Team).offset(offset).limit(limit)).all()
    return teams


@app.get("/teams/{team_id}", response_model=TeamPublicWithHeroes)
def read_team(*, team_id: int, session: Session = Depends(get_session)):
    team = session.get(Team, team_id)
    if not team:
        raise HTTPException(status_code=404, detail="Team not found")
    return team


@app.patch("/teams/{team_id}", response_model=TeamPublic)
def update_team(
    *,
    session: Session = Depends(get_session),
    team_id: int,
    team: TeamUpdate,
):
    db_team = session.get(Team, team_id)
    if not db_team:
        raise HTTPException(status_code=404, detail="Team not found")
    team_data = team.model_dump(exclude_unset=True)
    for key, value in team_data.items():
        setattr(db_team, key, value)
    session.add(db_team)
    session.commit()
    session.refresh(db_team)
    return db_team


@app.delete("/teams/{team_id}")
def delete_team(*, session: Session = Depends(get_session), team_id: int):
    team = session.get(Team, team_id)
    if not team:
        raise HTTPException(status_code=404, detail="Team not found")
    session.delete(team)
    session.commit()
    return {"ok": True}
```

### Description

I am trying to work out what the best method is for creating SQLModel objects with nested objects beneath them.
The example code in the tutorial shows reading data (such as with HeroPublicWithTeam) but not creating data.
My code above seems to work ok - it is just the `Models with Relationships in FastAPI` code with the HeroCreate class modified to optionally allow a Team to also be created within the same API call - and team_id moved from HeroBase to Hero as it doesn't make sense for HeroCreate to have a team and a team_id in it.

Is this the best way to achieve this?
Or am I missing a simpler way to do this that doesn't involve an AfterValidator which seems like an odd way to achieve this?
Just having `team: TeamCreate | None = None` leads to `'TeamCreate' object has no attribute '_sa_instance_state'`
Just having `team: Team | None = None` means that `id` is part of the API for creating a team within the hero creation leading to the inevitable `sqlalchemy.exc.IntegrityError: (psycopg.errors.UniqueViolation) duplicate key value violates unique constraint "team_pkey"`

### Operating System

Linux

### Operating System Details

python-slim docker container

### SQLModel Version

0.0.22

### Python Version

3.12.6

**Comment by shuryhin-oleksandr** [comment]
Hi, @jonyscathe I tried your solution and it also worked for me, but there is still a thing to consider. 

At the first line of a path operation, when fastapi validation is finished and before `Hero.model_validate(hero)` is invoked, what sort of data lives in hero?

Hero itself has a `HeroCreate` type, and hero.team has a `Team` type. So we end up with an object, where one part is a usual data model, and another part is a table model.

I am not sure where, but I have an intuition, that this inconsistency could cause a bug down a stream.

I was thinking to write a `@model_validator` for each data model, so I have table models right after fastapi path operation validations are finished. That still introduces some coupling + in official docs `.model_validate()` is called by hands. 

Another option is to not perform any automatic conversions and write all by hands. This approach is recommended in DRF, for example:
> Because the behavior of nested creates and updates can be ambiguous, and may require complex dependencies between related models, REST framework 3 requires you to always write these methods explicitly.

At last, what is a need to perform an object creation in path operation code? This data is more likely to be passed to some service containing a business logic, which handles connected objects creation.

So for myself I would either call pop teams info and call `.model_validate()` for Hero and Team separately, either pass this data to business logic layer as is.

PS: another catch - if you pass a foreign key of non-existing key while creating a book - it will return 500 instead of 400 with a fancy error description, because 666 is still a valid int, although team with id 666 does not exist.

**Comment by gzu300** [comment, marked as answer]
I ran into the same issue and was inspired by this thread. Here’s my refinement:

I introduced **Helper models** that sit between the `Create` schemas and the `table=True` ORM models. These helpers use `@field_validator(..., mode="after")` to recursively convert nested data into ORM-compatible objects — without mutating the original input schemas or embedding transformation logic into the ORM models.

This approach keeps input validation clean while making the transformation layer explicit, type-safe, and scalable.

### ✅ Minimal working example

```python
from sqlmodel import SQLModel, Field, Relationship
from pydantic import field_validator

# --- Input Schemas ---
class AddressCreate(SQLModel):
    city: str

class UserCreate(SQLModel):
    name: str
    address: AddressCreate

# --- ORM Models ---
class AddressModel(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    city: str

class UserModel(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    name: str
    address_id: int | None = Field(default=None, foreign_key="addressmodel.id")
    address: AddressModel = Relationship()

# --- Helper Models ---
class AddressCreateHelper(AddressCreate):
    @field_validator("city", mode="after")
    @classmethod
    def to_model(cls, v, values):
        return AddressModel(city=v)

class UserCreateHelper(UserCreate):
    @field_validator("address", mode="after")
    @classmethod
    def to_model(cls, v):
        return AddressModel.model_validate(AddressCreateHelper.model_validate(v))

# --- Usage ---
payload = {"name": "Alice", "address": {"city": "Berlin"}}

# In the FastAPI route or controller layer
user_input = UserCreate.model_validate(payload)

# In the transformation layer (e.g., service or use case layer)
user_helper = UserCreateHelper.model_validate(user_input)

# In the persistence layer
user_model = UserModel.model_validate(user_helper)
```

This keeps `UserCreate`, `UserModel`, and `UserCreateHelper` cleanly separated for API input, database persistence, and transformation logic — making the system easier to maintain and reason about.

**Reply by jonyscathe** [reply]
Nice solution. I ended up doing something kinda similar (though not as neat) and never updated this thread so I am glad someone did.

**Reply by space-romn** [reply]
Hi @gzu300 , would you like to share the rest of the script with the _path operations_?
How do you manage to create all the database object from the nested payload?
Thank you.

**Reply by erikdao** [reply]
Thanks for the nice solution! Definitely a helpful reference for the similar problem I'm facing.
